server/data/data-api/src/main/java/org/apache/james/user/api/UsersRepository.java

```diff
@@ -142,4 +142,10 @@
      */
     boolean isAdministrator(String username) throws UsersRepositoryException;
 
+    /**
+     * @return true if one can use {@link UsersRepository#updateUser(User)} {@link UsersRepository#addUser(String, String)}
```

Comment:
You are arguing runtime vs compile-time (or startup-time).
The later prevent more errors then the former, that's why I always prefer it.
Think about it with another example (from mutable programming world): a function take a collection, I give you an ImmutableList. If you call `add` it fails. But everything is find until you reach this piece of code.
You will never be able to enforce any piece of code to call isReadOnly before calling other methods. Why provide these method in the first place if you can't rely on them ?

Splitting ReadOnlyUsersRepository from ReadWriteUsersRepository would make it very clear that if your application need to add users, you can't use LDAP as an implementation.

We would not have run into current troubles if it were designed like that in the first place, isn't it ?